File: nodejs-mvc/layered-architecture/src/presentation/controllers/ProductController.ts
import { Router, Request, Response, NextFunction } from 'express';
import { ProductService } from '../../business/services/ProductService';
import {
  CreateProductRequest,
  ProductResponse,
  UpdateProductRequest,
  PaginatedResponse,
} from '../dto';
import { Product } from '../../data/models/Product';
import { ConflictError, InsufficientStockError, NotFoundError, ValidationError } from '../../business/errors';

/**
 * @layer Presentation Layer
 * @description Controller that handles HTTP requests and delegates to the Business layer.
 * In Layered Architecture, the Presentation layer is the topmost layer.
 *
 * Dependencies flow DOWNWARD:
 *   Presentation → Business → Data
 *
 * The controller:
 * 1. Parses HTTP input
 * 2. Calls the business service
 * 3. Maps domain objects to response DTOs
 * 4. Handles errors and sets HTTP status codes
 */
export class ProductController {
  public readonly router: Router;

  constructor(private readonly productService: ProductService) {
    this.router = Router();
    this.setupRoutes();
  }

  private setupRoutes(): void {
    this.router.post('/', this.createProduct.bind(this));
    this.router.get('/search', this.searchProducts.bind(this));
    this.router.get('/', this.listProducts.bind(this));
    this.router.get('/:id', this.getProductById.bind(this));
    this.router.put('/:id', this.updateProduct.bind(this));
    this.router.delete('/:id', this.deleteProduct.bind(this));
    this.router.patch('/:id/stock/decrease', this.decreaseStock.bind(this));
    this.router.patch('/:id/stock/increase', this.increaseStock.bind(this));
  }

  private async createProduct(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const body: CreateProductRequest = req.body;

      if (!body.name || !body.description || !body.price || !body.category || !body.sku) {
        res.status(400).json({
          error: 'Bad Request',
          message: 'Missing required fields: name, description, price, category, sku',
          timestamp: new Date().toISOString(),
        });
        return;
      }

      if (body.stockQuantity === undefined || body.stockQuantity === null) {
        res.status(400).json({
          error: 'Bad Request',
          message: 'Missing required field: stockQuantity',
          timestamp: new Date().toISOString(),
        });
        return;
      }

      const product = await this.productService.createProduct(body);
      res.status(201).json(this.toResponse(product));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async listProducts(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const page = parseInt(req.query.page as string) || 0;
      const size = parseInt(req.query.size as string) || 20;

      const result = await this.productService.listProducts(page, size);

      const response: PaginatedResponse<ProductResponse> = {
        content: result.content.map(this.toResponse),
        page: result.page,
        size: result.size,
        totalElements: result.totalElements,
        totalPages: result.totalPages,
      };

      res.json(response);
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async getProductById(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const product = await this.productService.getProductById(req.params.id);
      res.json(this.toResponse(product));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async searchProducts(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { query, category, minPrice, maxPrice } = req.query;

      const products = await this.productService.searchProducts(
        query as string,
        category as string,
        minPrice ? parseFloat(minPrice as string) : undefined,
        maxPrice ? parseFloat(maxPrice as string) : undefined
      );

      res.json(products.map(this.toResponse));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async updateProduct(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const body: UpdateProductRequest = req.body;
      const product = await this.productService.updateProduct(req.params.id, body);
      res.json(this.toResponse(product));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async deleteProduct(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      await this.productService.deleteProduct(req.params.id);
      res.status(204).send();
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async decreaseStock(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const quantity = parseInt(req.query.quantity as string);
      if (!quantity || quantity <= 0) {
        res.status(400).json({
          error: 'Bad Request',
          message: 'Quantity must be a positive number',
          timestamp: new Date().toISOString(),
        });
        return;
      }

      const product = await this.productService.decreaseStock(req.params.id, quantity);
      res.json(this.toResponse(product));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private async increaseStock(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const quantity = parseInt(req.query.quantity as string);
      if (!quantity || quantity <= 0) {
        res.status(400).json({
          error: 'Bad Request',
          message: 'Quantity must be a positive number',
          timestamp: new Date().toISOString(),
        });
        return;
      }

      const product = await this.productService.increaseStock(req.params.id, quantity);
      res.json(this.toResponse(product));
    } catch (error) {
      this.handleError(error, res);
    }
  }

  private toResponse(product: Product): ProductResponse {
    return {
      id: product.id,
      name: product.name,
      description: product.description,
      price: product.price,
      category: product.category,
      stockQuantity: product.stockQuantity,
      sku: product.sku,
      active: product.active,
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };
  }

  private handleError(error: unknown, res: Response): void {
    if (error instanceof NotFoundError) {
      res.status(404).json({ error: 'Not Found', message: error.message, timestamp: new Date().toISOString() });
    } else if (error instanceof ValidationError || error instanceof InsufficientStockError) {
      res.status(400).json({ error: 'Bad Request', message: error.message, timestamp: new Date().toISOString() });
    } else if (error instanceof ConflictError) {
      res.status(409).json({ error: 'Conflict', message: error.message, timestamp: new Date().toISOString() });
    } else {
      res.status(500).json({ error: 'Internal Server Error', message: 'An unexpected error occurred', timestamp: new Date().toISOString() });
    }
  }
}

File: python-mvc/layered-architecture/app/presentation/product_controller.py
"""
Presentation Layer: Product REST Controller.

FastAPI router that handles HTTP requests, validates input via DTOs,
delegates to the Business Layer, and maps errors to HTTP status codes.

LAYERED ARCHITECTURE FLOW:
    HTTP Request → Controller (Presentation) → Service (Business) → Repository (Data)
    HTTP Response ← Controller (Presentation) ← Service (Business) ← Repository (Data)

KEY DIFFERENCE FROM HEXAGONAL ARCHITECTURE:
- Hexagonal: The controller (input adapter) depends on an abstract input port.
  It never imports the concrete service. Error types come from the domain core.
- Layered: The controller directly imports and depends on the concrete
  ProductService class and its error types from the Business Layer.
  There is no indirection through ports/interfaces.

This direct dependency makes the code simpler to follow but means the
Presentation Layer is tightly coupled to the Business Layer's implementation.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, status

from app.business.errors import (
    ConflictError,
    InsufficientStockError,
    NotFoundError,
    ValidationError,
)
from app.business.product_service import ProductService
from app.presentation.dto import (
    CreateProductRequest,
    PaginatedResponse,
    ProductResponse,
    UpdateProductRequest,
)

router = APIRouter(prefix="/api/v1/products", tags=["Products"])

# Service instance — wired at application startup
_product_service: ProductService | None = None


def set_product_service(service: ProductService) -> None:
    """Wire the product service into the controller at startup."""
    global _product_service
    _product_service = service


def _get_service() -> ProductService:
    """Get the product service instance."""
    if _product_service is None:
        raise RuntimeError("ProductService has not been initialized")
    return _product_service


def _to_response(product) -> ProductResponse:
    """Map a Product data model to a ProductResponse DTO."""
    return ProductResponse(
        id=product.id,
        name=product.name,
        description=product.description,
        price=product.price,
        category=product.category,
        stock_quantity=product.stock_quantity,
        sku=product.sku,
        active=product.active,
        created_at=product.created_at,
        updated_at=product.updated_at,
    )


@router.post(
    "",
    response_model=ProductResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_product(request: CreateProductRequest) -> ProductResponse:
    """Create a new product in the catalog."""
    try:
        product = await _get_service().create_product(
            name=request.name,
            description=request.description,
            price=request.price,
            category=request.category,
            stock_quantity=request.stock_quantity,
            sku=request.sku,
        )
        return _to_response(product)
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )
    except ConflictError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail=str(e)
        )


@router.get("", response_model=PaginatedResponse)
async def list_products(
    page: int = Query(1, ge=1, description="Page number"),
    size: int = Query(10, ge=1, le=100, description="Page size"),
) -> PaginatedResponse:
    """List all products with pagination."""
    result = await _get_service().list_products(page=page, size=size)
    return PaginatedResponse(
        items=[_to_response(p) for p in result["items"]],
        total=result["total"],
        page=result["page"],
        size=result["size"],
        total_pages=result["total_pages"],
    )


@router.get("/search", response_model=list[ProductResponse])
async def search_products(
    query: str | None = Query(None, description="Search query for name/description"),
    category: str | None = Query(None, description="Filter by category"),
    min_price: float | None = Query(None, ge=0, description="Minimum price"),
    max_price: float | None = Query(None, ge=0, description="Maximum price"),
) -> list[ProductResponse]:
    """Search products by various criteria."""
    products = await _get_service().search_products(
        query=query,
        category=category,
        min_price=min_price,
        max_price=max_price,
    )
    return [_to_response(p) for p in products]


@router.get("/{product_id}", response_model=ProductResponse)
async def get_product(product_id: str) -> ProductResponse:
    """Get a product by its ID."""
    try:
        product = await _get_service().get_product_by_id(product_id)
        return _to_response(product)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )


@router.put("/{product_id}", response_model=ProductResponse)
async def update_product(
    product_id: str, request: UpdateProductRequest
) -> ProductResponse:
    """Update an existing product."""
    try:
        product = await _get_service().update_product(
            product_id=product_id,
            name=request.name,
            description=request.description,
            price=request.price,
            category=request.category,
            stock_quantity=request.stock_quantity,
            sku=request.sku,
            active=request.active,
        )
        return _to_response(product)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )
    except ConflictError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail=str(e)
        )


@router.delete("/{product_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_product(product_id: str) -> None:
    """Delete a product by its ID."""
    try:
        await _get_service().delete_product(product_id)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )


@router.patch("/{product_id}/stock/decrease", response_model=ProductResponse)
async def decrease_stock(
    product_id: str,
    quantity: int = Query(..., gt=0, description="Quantity to decrease"),
) -> ProductResponse:
    """Decrease the stock of a product."""
    try:
        product = await _get_service().decrease_stock(product_id, quantity)
        return _to_response(product)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )
    except (InsufficientStockError, ValidationError) as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )


@router.patch("/{product_id}/stock/increase", response_model=ProductResponse)
async def increase_stock(
    product_id: str,
    quantity: int = Query(..., gt=0, description="Quantity to increase"),
) -> ProductResponse:
    """Increase the stock of a product."""
    try:
        product = await _get_service().increase_stock(product_id, quantity)
        return _to_response(product)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )
